### PR TITLE
Replace pimoroni trackball driver.

### PR DESCRIPTION
**Pull request by daskygit**
## Description

WIP

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/pointing_device/pointing_device_drivers.c [hunk 1]
```diff
@@ -268,24 +269,64 @@ mouse_xy_report_t pimoroni_trackball_adapt_values(clamp_range_t* offset) {
...
+                            began_motion += last_read_elapsed * 4;
+                            if (began_motion > last_read) {
+                                began_motion = last_read;
+                            }
+                        }
+                    }
+
+                    if (in_motion) {
+                        mouse_report.x = (int8_t)pimoroni_data.left - (int8_t)pimoroni_data.right;
+                        mouse_report.y = (int8_t)pimoroni_data.up - (int8_t)pimoroni_data.down;
```
Comment (drashna) [inline quantum/pointing_device/pointing_device_drivers.c:309]: x and y can be int16_t.  `mouse_xy_report_t` is typedef'd to the correct range, and may be more appropriate here. 
```suggestion
                        mouse_report.x = (mouse_xy_report_t)pimoroni_data.left - (mouse_xy_report_t)pimoroni_data.right;
                        mouse_report.y = (mouse_xy_report_t)pimoroni_data.up - (mouse_xy_report_t)pimoroni_data.down;
```

**Review by drashna — APPROVED** [review]

File: quantum/pointing_device/pointing_device_drivers.c (hunk 1 above)
Comment (drashna) [reply quantum/pointing_device/pointing_device_drivers.c:309]: Yeah, figured as much.

**Review by keyboard-magpie — APPROVED** [review]
Have been testing this this evening. Seems to handle extremes of motion better than previous drivers. Not found any issues.

**Comment by BenRoe** [comment]
Tested it and it runs smoothly. Only problem is, the cursor moves so slow/less even if i turn up the settings in the Mac System Mouse preferences.
Any solution for that?

**Comment by daskygit** [comment]
> Tested it and it runs smoothly. Only problem is, the cursor moves so slow/less even if i turn up the settings in the Mac System Mouse preferences. Any solution for that?

@BenRoe I haven't pushed this any further as the driver behaves differently depending on how frequently the trackball is polled. A solution could be to limit the polling rate with `#define POINTING_DEVICE_TASK_THROTTLE_MS 8` in config.h, would you mind testing this? It should accelerate with quick repeatable trackball movements.

**Comment by mbissonn** [comment]
Hi! This has completely changed my experience with the Pimoroni. The only downside is that I seem to have lost its LED functionality. Previously, I was using:

void keyboard_post_init_user(void) {
    pimoroni_trackball_set_rgbw(0,0,0,100);
}

That code doesn't seem to work with your driver. Is there another way that I can talk to the LED or something that I am missing? Thanks!

**Comment by daskygit** [comment]
It should still be working, I did add a check so it'll only write if there's been a successful read. I guess there might be a condition on boot where the i2c read isn't successful. You mind trying this instead as a test? @mbissonn 

```c
void housekeeping_task_user(void) {
    pimoroni_trackball_set_rgbw(0,0,0,100);
}
```
The above code will run every loop.

**Comment by mbissonn** [comment]
No luck with the LED, but running the command every loop did noticeably decrease the smoothness for me.

**Comment by daskygit** [comment]
> No luck with the LED, but running the command every loop did noticeably decrease the smoothness for me.

Do you have discord? Probably better to do the troubleshooting on there if we can. https://discord.com/invite/Uq7gcHh

**Comment by lucacri** [comment]
I just tested this PR, and in order to use it with the rebased version of QMK it requires one change in line 80 of `pimoroni_trackball.c`:

`        pd_dprintf("Trackball READ i2c_status_t: %d L: %d R: %d Up: %d D: %d SW: %d\n", last_i2c_status, data->left, data->right, data->up, data->down, data->click);`

Gotta say, it behaves MUCH better when using it in combination with `#define POINTING_DEVICE_TASK_THROTTLE_MS 8`.

Now my question is.. why is this not merged? Is there a problem/bug in it? I am not so well versed in this device but it looks good to me

**Comment by drashna** [comment]
Mostly, it's just waiting on dasky to confirm that it's ready to go.

**Comment by daskygit** [comment]
I apologise for the delay, this has sat far too long now. I will be closing this PR without merge.

The expected behaviour only really works with a polling rate of 8ms as polling at faster rates reduces to the movement deltas to 1-2. I'll probably try for a simpler driver again at some point in the future.

**Pull request closed without merging**